src/daft-logical-plan/src/optimization/rules/extract_window_function.rs

```diff
@@ -0,0 +1,488 @@
+
+        Ok(transformed.data)
+    }
+}
+
+impl OptimizerRule for ExtractWindowFunction {
+    /// Optimizes a logical plan by extracting window function expressions from projections
+    /// and converting them into explicit Window operations.
+    ///
+    /// Algorithm:
+    /// 1. Find all window functions in a projection
+    /// 2. Group them by their window specifications
+    /// 3. Create a separate Window operation for each group
+    /// 4. Connect these Window operations in a chain
+    /// 5. Replace the original window expressions with column references
+    /// 6. Create a final Project operation with the updated expressions
+    fn try_optimize(&self, plan: Arc<LogicalPlan>) -> DaftResult<Transformed<Arc<LogicalPlan>>> {
+        plan.transform_down(|node| match node.as_ref() {
+            LogicalPlan::Project(project) => {
+                if Self::contains_window_function(project) {
+                    let window_funcs = Self::extract_window_functions(&project.projection);
+
+                    debug_assert!(
+                        !window_funcs.is_empty(),
+                        "Window functions detected but none extracted"
+                    );
+
+                    let window_funcs_grouped_by_spec = window_funcs
+                        .into_iter()
+                        .into_group_map_by(|(_, spec)| spec.clone());
+
+                    let mut window_col_mappings = Vec::new();
+
+                    let mut current_plan = project.input.clone();
+
+                    for (window_spec, window_exprs_for_spec) in window_funcs_grouped_by_spec {
+                        let window_function_exprs = window_exprs_for_spec
+                            .iter()
+                            .map(|(expr, _)| expr.clone())
+                            .collect::<Vec<ExprRef>>();
+
+                        let window_function_exprs_aliased = window_function_exprs
+                            .iter()
+                            .map(|e| {
+                                let semantic_id = e.semantic_id(&current_plan.schema());
+                                e.alias(semantic_id.id)
+                            })
+                            .collect::<Vec<ExprRef>>();
+
+                        current_plan = Arc::new(LogicalPlan::Window(Window::try_new(
+                            current_plan,
+                            window_function_exprs_aliased,
+                            window_spec,
+                        )?));
+
+                        let spec_mappings = window_function_exprs
+                            .iter()
+                            .map(|e| {
+                                let semantic_id = e.semantic_id(&current_plan.schema());
+                                (e.clone(), semantic_id.id.to_string())
+                            })
+                            .collect::<Vec<(ExprRef, String)>>();
+
+                        window_col_mappings.extend(spec_mappings);
+                    }
```

Comment:
tiny nit: you could use [Iterator::fold](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold) here instead for more idiomatic rust